src/happy/oauth2.clj

```diff
@@ -6,64 +6,61 @@
-  (:import (java.util Date)))
+  (:import (java.util Date)
+           (java.net URLEncoder)))
 
-(defn- redirect-to-google
-  "Step 2: Redirect to Google's OAuth 2.0 server.
+(set! *warn-on-reflection* true)
+
+(defn url-encode [^String s]
+  (URLEncoder/encode s "UTF-8"))
+
+(defn query-string
+  "If you have clj-http as a dependency, prefer http.client/generate-query-string instead."
+  [query-params]
+  (str/join "&" (for [[k v] query-params]
+                  (str (url-encode (name k)) "=" (url-encode (str v))))))
+
+(defn provider-login-url
+  "Step 1: Set authorization parameters.
   Builds the URL to send the user to for them to authorize your app.
-  Do not call this function directly, it is called from set-authorization-parameters (step 1).
   For local testing you can paste this URL into your browser,
   or call (clojure.java.browse/browse-url (redirect-to-google my-config)).
   In your app you need to send your user to this URL, usually with a redirect response."
-  [{:as config :keys [client_id redirect_uri redirect_uris]} scopes {:as optional :keys [state login_hint]}]
-  {:pre [client_id (or redirect_uri redirect_uris) scopes]}
-  (str "https://accounts.google.com/o/oauth2/v2/auth"
-       "?client_id=" client_id
-       "&state=" state
-       "&response_type=code"
-       "&access_type=offline"
-       "&scope=" (str/join "%20" scopes)
-       "&redirect_uri=" (or redirect_uri (last redirect_uris))))
-
-(defn set-authorization-parameters
-  "Step 1: Set authorization parameters.
-  We tell Google to expect a visit from our user.
-  Returns the URL to redirect the user to."
-  ([config scopes] (set-authorization-parameters config scopes nil))
-  ([{:as config :keys [client_id redirect_uri redirect_uris]} scopes {:as optional :keys [state login_hint]}]
-   {:pre [client_id (or redirect_uri redirect_uris) scopes]}
-   (assert
-     (= 200
-        (:status
-          (http/get "https://accounts.google.com/o/oauth2/v2/auth"
-                    {:query-params (merge {:client_id     client_id
-                                           :redirect_uri  (or redirect_uri (last redirect_uris))
-                                           :response_type "code"
-                                           :scope         (str/join " " scopes)}
-                                          optional)}))))
-   (redirect-to-google config scopes optional)))
+  [{:as config :keys [auth_uri client_id redirect_uri redirect_uris]} scopes {:as optional :keys [state login_hint]}]
+  {:pre [auth_uri client_id (or redirect_uri redirect_uris) scopes]}
+  (str auth_uri "?"
+       (query-string (merge {:client_id     client_id
+                             :response_type "code"
+                             :redirect_uri  (or redirect_uri (last redirect_uris))
+                             :access_type   "offline"
+                             :scope         (str/join " " scopes)}
+                            optional))))
```

Comment:
Good ideas;
Right - previously I was only selecting some params from optional, but that didn't seem necessary and prevented some options uncessarily.
I'm adding links to the optional documentation and adding a default `state` param seeing it is recommended. Also I'm moving the access_type offline default to fresh-credentials (because those sort of defaults should probably only exist for the reference implementation of credentials) and adding include_granted_scopes true as that seems like a useful thing.